uktrade/data-hub-frontend
# Add sagas to handle flash messages in session storage

## Description of change

This PR updates the way that the frontend interacts with session storage in order to pass flash messages between pages.

### The problem we're solving

Previously, the `FlashMessages` component got messages directly from the session and was also responsible for clearing them. This caused issues because when submitting a form on a page which used the react local header (containing `<FlashMessages>`), clearMessages() would be called as the local header is re-rendered before redirecting, thus removing the flash message from session storage before it was rendered on the next page.

### Our solution

This PR fixes this issue, by using redux sagas to handle side effects - ie. interactions with our session storage.

One saga takes an action dispatched on successful submission of the form and adds the message from the body of that action to the session storage.

The other saga awaits an action to be dispatched (which happens on render of the FlashMessage component within LocalHeader), then dispatches an action to add the message from session storage into the state.

To account for when a redirect is made via react router (so there is no page refresh), the useEffect hook which dispatches the action in the FlashMessages component has a cleanup method which clears the flashMessages section of state when the component is unmounted.

### TO NOTE Testing

We have not added any additional tests in this PR, as the expected behaviour of the flash message component hasn't changed. The existing functional tests check that messages appear after form submissions, but we have no tests to check the implementation detail. 

Any thoughts on whether we should add tests to check our implementation would be greatly appreciated! 

## Test instructions

No visual changes.

You can check the behaviour of the flash messages when redirecting with react router by:
1. Editing an event
2. Checking the flash message appears on the event details page
3. Click the 'Edit event' button
4. Check the flash message no longer appears in the local header

## Checklist

[//]: # "When submitting a PR make sure the code review guidelines have been satisfied.
https://github.com/uktrade/data-hub-frontend/blob/master/docs/Code%20review%20guidelines.md"
